src/openms/source/ANALYSIS/ID/SearchDatabase.cpp

```diff
@@ -0,0 +1,337 @@
+      all_frags.emplace_back(i, frag);        
+    }
+    chunk_start.emplace_back(all_frags.size());
+    b_y_ions.clear(true);
+  }
+
+  SearchDatabase::fragment_merge_(0, chunk_start.size()-1, chunk_start, all_frags);
+  
+  return all_frags;
+}
+
+SearchDatabase::SearchDatabase(const std::vector<FASTAFile::FASTAEntry>& entries) : DefaultParamHandler("SearchDatabase")
+{ 
+  vector<String> all_enzymes;  
+  ProteaseDB::getInstance()->getAllNames(all_enzymes);
+  vector<String> all_mods;
+  ModificationsDB::getInstance()->getAllSearchModifications(all_mods);
+  vector<string> tolerance_units{UNIT_DA, UNIT_PPM};
+  defaults_.setValue("digestor_enzyme", "Trypsin", "Enzyme for digestion");
+  defaults_.setValidStrings("digestor_enzyme", ListUtils::create<std::string>(all_enzymes));
+  defaults_.setValue("missed_cleavages", 0, "Missed cleavages for digestion");
+  defaults_.setValue("peptide_min_mass", 500, "Minimal peptide mass for database");
+  defaults_.setValue("peptide_max_mass", 5000, "Maximal peptide mass for database");
+  defaults_.setValue("peptide_min_length", 5, "Minimal peptide length for database");
+  defaults_.setValue("peptide_max_length", 50, "Maximal peptide length for database");
+  defaults_.setValue("fragment_min_mz", 150, "Minimal fragment mz for database");
+  defaults_.setValue("fragment_max_mz", 2000, "Maximal fragment mz for database");
+  defaults_.setValue("precursor_mz_tolerance", 2.0, "Tolerance for precursor-m/z in search");
+  defaults_.setValue("fragment_mz_tolerance", 0.05, "Tolerance for fragment-m/z in search");
+  defaults_.setValue("precursor_mz_tolerance_unit", UNIT_DA, "Unit of tolerance for precursor-m/z");
+  defaults_.setValidStrings("precursor_mz_tolerance_unit", tolerance_units);
+  defaults_.setValue("fragment_mz_tolerance_unit", UNIT_DA, "Unit of tolerance for fragment-m/z");
+  defaults_.setValidStrings("fragment_mz_tolerance_unit", tolerance_units);
+  defaults_.setValue("modifications_fixed", std::vector<std::string>{"Carbamidomethyl (C)"}, "Fixed modifications, specified using UniMod (www.unimod.org) terms, e.g. 'Carbamidomethyl (C)'");
+  defaults_.setValidStrings("modifications_fixed", ListUtils::create<std::string>(all_mods));
+  defaults_.setValue("modifications_variable", std::vector<std::string>{"Oxidation (M)"}, "Variable modifications, specified using UniMod (www.unimod.org) terms, e.g. 'Oxidation (M)'");
+  defaults_.setValidStrings("modifications_variable", ListUtils::create<std::string>(all_mods));
+  defaults_.setValue("max_variable_mods_per_peptide", 2, "Maximum number of residues carrying a variable modification per candidate peptide");  
+
+  defaultsToParam_();
+
+  if (entries.size() == 0) return;
+
+  all_peptides_ = generate_peptides_(entries);
+
+  all_fragments_ = generate_fragments_();
+  
+  bucketsize_ = size_t(sqrt(all_fragments_.size())); //calculating bucketsize for balanced tree (performance)
+
+  for (size_t i = 0; i < all_fragments_.size(); i += bucketsize_)
```

Comment:
add some comment what is calculated here. e.g., store lower m/z border of buckets